{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Neale diseases genes (Pascal) and diseases genes (GWAS catalog)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### First version 17/06/2022\n",
    "###### Latest modif 18/10/2022"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "import seaborn as sns\n",
    "from datetime import datetime\n",
    "import os\n",
    "import glob\n",
    "\n",
    "DATE = datetime.now().strftime(\"%Y-%m-%d\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save_dir ='/SSD/home/sofia/retina-phenotypes/complementary/GWAS_postprocessing/common_genes_path_snps/'\n",
    "save_dir ='/NVME/decrypted/scratch/multitrait/UK_BIOBANK_ZERO/genes/'\n",
    "df_genes_count= pd.read_csv(save_dir +'/genes_count_main_v5.csv')\n",
    "\n",
    "neale_csv_name='11_10_2022_diseases_genes.csv'\n",
    "#df_genes_count.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Pre analysis: (example) Show Neale diseases Pascal genes for SBP and DBP:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_ = '/NVME/decrypted/scratch/multitrait/UK_BIOBANK_ZERO/gwas/2022_08_03_ventile5/gcorr_diseases/'\n",
    "file_4079='4079_irnt.gwas.imputed_v3.both_sexes.tsv__gene_scores.p'\n",
    "\n",
    "df_4079=pd.read_pickle(dir_+file_4079)\n",
    "\n",
    "file_4080='4080_irnt.gwas.imputed_v3.both_sexes.tsv__gene_scores.p'\n",
    "\n",
    "df_4080=pd.read_pickle(dir_+file_4080)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"[[], [], [['MIR1302-10', 'No SNPs'], ['FAM138A', 'No SNPs'], ['OR4F5', 'No SNPs'], ['RP11-34P13.7', 'No SNPs'], ['RP11-34P13.8', 'No SNPs'], ['AL627309.1', 'No SNPs'], ['RP11-34P13.9', 'No SNPs'], ['AP006222.2', 'No SNPs'], ['RP4-669L17.10', 'No SNPs'], ['OR4F29', 'No SNPs'], ['RP4-669L17.2', 'No SNPs'], ['RP5-857K21.15', 'No SNPs'], ['RP5-857K21.1', 'No SNPs'], ['RP5-857K21.2', 'No SNPs'], ['RP5-857K21.3', 'No SNPs'], ['RP5-857K21.4', 'No SNPs'], ['RP5-857K21.5', 'No SNPs'], ['OR4F16', 'No SNPs'], ['RP11-206L10.3', 'No SNPs'], ['RP11-206L10.5', 'No SNPs'], ['RP11-206L10.2', 'No SNPs'], ['RP11-206L10.9', 'No SNPs'], ['AL669831.1', 'No SNPs'], ['FAM87B', 'No SNPs'], ['LINC00115', 'No SNPs'], ['FAM41C', 'No SNPs'], ['AL645608.2', 'No SNPs'], ['RP11-54O7.16', 'No SNPs'], ['RP11-54O7.1', 'No SNPs'], ['RP11-54O7.2', 'No SNPs'], ['RP11-54O7.3', 'No SNPs'], ['SAMD11', 'No SNPs'], ['AL645608.1', 'No SNPs'], ['NOC2L', 'No SNPs'], ['KLHL17', 'No SNPs'], ['PLEKHN1', 'No SNPs'], ['C1orf170', 'No S\""
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(df_4080)[0:1000]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1- Read the Neale Pascal genes for all the diseases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_value_min = 5.7  # -math.log10(0.05/len(x))\n",
    "\n",
    "#input_dir='/HDD/data/ukbb/disease_sumstats/files_modified/'\n",
    "input_dir='/HDD/data/ukbb/disease_sumstats/VARIANTS/'\n",
    "l_diseases_all=[]\n",
    "\n",
    "#os.chdir('/HDD/data/ukbb/disease_sumstats/files_modified/')\n",
    "os.chdir('/HDD/data/ukbb/disease_sumstats/VARIANTS/')\n",
    "for file in glob.glob(\"*__gene_scores\"):\n",
    "    #print(file)\n",
    "    #file='4080_irnt.gwas.imputed_v3.both_sexes.tsv'\n",
    "    #df = pd.read_csv(input_dir+file+'__gene_scores', delimiter='\\t', names =['gen', 'p']) #, index_col=None, header=0)\n",
    "    df = pd.read_csv(input_dir+file, delimiter='\\t', names =['gen', 'p']) #, index_col=None, header=0)\n",
    "\n",
    "    df['file_col']=file\n",
    "\n",
    "    df['-log10(p)'] = -np.log10(df['p'])\n",
    "    y = df[df['-log10(p)'] >= p_value_min]\n",
    "    df_significant = y.sort_values('-log10(p)', ascending=False)\n",
    "    pheno, b = file.split(\".gwas.imputed_v3.both_sexes.tsv__gene_scores\")\n",
    "    aux, pheno_name = pheno.split(\"only_rs_notna_mod_variants_\")\n",
    "    \n",
    "    data={\n",
    "    'pheno':  pheno,\n",
    "    'name_pheno': pheno_name,\n",
    "    'N genes': len(df_significant),\n",
    "    'names': df_significant['gen'].to_list()\n",
    "    }\n",
    "    l_diseases_all.append(data)\n",
    "\n",
    "df_diseases_all =pd.DataFrame(l_diseases_all)\n",
    "    \n",
    "    #print(pheno, ' Significant number of genes: ', len(df_significant))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pheno</th>\n",
       "      <th>name_pheno</th>\n",
       "      <th>N genes</th>\n",
       "      <th>names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>only_rs_notna_mod_variants_5843_1</td>\n",
       "      <td>5843_1</td>\n",
       "      <td>1</td>\n",
       "      <td>[EXOSC4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>only_rs_notna_mod_variants_1717</td>\n",
       "      <td>1717</td>\n",
       "      <td>468</td>\n",
       "      <td>[CDK10, RP11-368I7.4, DPEP1, CHMP1A, SPATA33, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>only_rs_notna_mod_variants_4080_irnt</td>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>959</td>\n",
       "      <td>[NPPA, NPPB, MTHFR, CLCN6, C1orf167, NPR3, AGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>only_rs_notna_mod_variants_5855_1</td>\n",
       "      <td>5855_1</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>only_rs_notna_mod_variants_4056_irnt</td>\n",
       "      <td>4056_irnt</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>only_rs_notna_mod_variants_4079_irnt</td>\n",
       "      <td>4079_irnt</td>\n",
       "      <td>910</td>\n",
       "      <td>[CSK, RP1-278O22.1, LMAN1L, CYP1A2, CPLX3, ULK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>only_rs_notna_mod_variants_4079_raw</td>\n",
       "      <td>4079_raw</td>\n",
       "      <td>905</td>\n",
       "      <td>[NPPA, NPPB, MTHFR, CLCN6, C1orf167, RP1-278O2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>only_rs_notna_mod_variants_3894_irnt</td>\n",
       "      <td>3894_irnt</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>only_rs_notna_mod_variants_5843_3</td>\n",
       "      <td>5843_3</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>only_rs_notna_mod_variants_3627_irnt</td>\n",
       "      <td>3627_irnt</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>only_rs_notna_mod_variants_102_raw</td>\n",
       "      <td>102_raw</td>\n",
       "      <td>884</td>\n",
       "      <td>[MYH6, CMTM5, IL25, EFS, KIAA1755, MYH7, SLC22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>only_rs_notna_mod_variants_21021_raw</td>\n",
       "      <td>21021_raw</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>only_rs_notna_mod_variants_21021_irnt</td>\n",
       "      <td>21021_irnt</td>\n",
       "      <td>3</td>\n",
       "      <td>[PALM, TEX41, PRSS57]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>only_rs_notna_mod_variants_5855_3</td>\n",
       "      <td>5855_3</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>only_rs_notna_mod_variants_5610_2</td>\n",
       "      <td>5610_2</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>only_rs_notna_mod_variants_40007_irnt</td>\n",
       "      <td>40007_irnt</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>only_rs_notna_mod_variants_4012_irnt</td>\n",
       "      <td>4012_irnt</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>only_rs_notna_mod_variants_5832_2</td>\n",
       "      <td>5832_2</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>only_rs_notna_mod_variants_1558</td>\n",
       "      <td>1558</td>\n",
       "      <td>276</td>\n",
       "      <td>[AC109829.1, ADH7, KLB, NPIPB8, SULT1A1, CCDC1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>only_rs_notna_mod_variants_102_irnt</td>\n",
       "      <td>102_irnt</td>\n",
       "      <td>907</td>\n",
       "      <td>[MYH6, CMTM5, IL25, EFS, KIAA1755, MYH7, SLC22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>only_rs_notna_mod_variants_4080_raw</td>\n",
       "      <td>4080_raw</td>\n",
       "      <td>838</td>\n",
       "      <td>[NPR3, C11orf89, LSP1, CACNB2, ATP2B1, TNNI2, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>only_rs_notna_mod_variants_4700_irnt</td>\n",
       "      <td>4700_irnt</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>only_rs_notna_mod_variants_4700_raw</td>\n",
       "      <td>4700_raw</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>only_rs_notna_mod_variants_2976_irnt</td>\n",
       "      <td>2976_irnt</td>\n",
       "      <td>106</td>\n",
       "      <td>[HLA-DQB1, XXbac-BPG254F23.7, HLA-DQA1, HLA-DQ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    pheno  name_pheno  N genes  \\\n",
       "0       only_rs_notna_mod_variants_5843_1      5843_1        1   \n",
       "1         only_rs_notna_mod_variants_1717        1717      468   \n",
       "2    only_rs_notna_mod_variants_4080_irnt   4080_irnt      959   \n",
       "3       only_rs_notna_mod_variants_5855_1      5855_1        0   \n",
       "4    only_rs_notna_mod_variants_4056_irnt   4056_irnt        0   \n",
       "5    only_rs_notna_mod_variants_4079_irnt   4079_irnt      910   \n",
       "6     only_rs_notna_mod_variants_4079_raw    4079_raw      905   \n",
       "7    only_rs_notna_mod_variants_3894_irnt   3894_irnt        0   \n",
       "8       only_rs_notna_mod_variants_5843_3      5843_3        0   \n",
       "9    only_rs_notna_mod_variants_3627_irnt   3627_irnt        0   \n",
       "10     only_rs_notna_mod_variants_102_raw     102_raw      884   \n",
       "11   only_rs_notna_mod_variants_21021_raw   21021_raw        0   \n",
       "12  only_rs_notna_mod_variants_21021_irnt  21021_irnt        3   \n",
       "13      only_rs_notna_mod_variants_5855_3      5855_3        0   \n",
       "14      only_rs_notna_mod_variants_5610_2      5610_2        0   \n",
       "15  only_rs_notna_mod_variants_40007_irnt  40007_irnt        0   \n",
       "16   only_rs_notna_mod_variants_4012_irnt   4012_irnt        0   \n",
       "17      only_rs_notna_mod_variants_5832_2      5832_2        0   \n",
       "18        only_rs_notna_mod_variants_1558        1558      276   \n",
       "19    only_rs_notna_mod_variants_102_irnt    102_irnt      907   \n",
       "20    only_rs_notna_mod_variants_4080_raw    4080_raw      838   \n",
       "21   only_rs_notna_mod_variants_4700_irnt   4700_irnt        0   \n",
       "22    only_rs_notna_mod_variants_4700_raw    4700_raw        0   \n",
       "23   only_rs_notna_mod_variants_2976_irnt   2976_irnt      106   \n",
       "\n",
       "                                                names  \n",
       "0                                            [EXOSC4]  \n",
       "1   [CDK10, RP11-368I7.4, DPEP1, CHMP1A, SPATA33, ...  \n",
       "2   [NPPA, NPPB, MTHFR, CLCN6, C1orf167, NPR3, AGT...  \n",
       "3                                                  []  \n",
       "4                                                  []  \n",
       "5   [CSK, RP1-278O22.1, LMAN1L, CYP1A2, CPLX3, ULK...  \n",
       "6   [NPPA, NPPB, MTHFR, CLCN6, C1orf167, RP1-278O2...  \n",
       "7                                                  []  \n",
       "8                                                  []  \n",
       "9                                                  []  \n",
       "10  [MYH6, CMTM5, IL25, EFS, KIAA1755, MYH7, SLC22...  \n",
       "11                                                 []  \n",
       "12                              [PALM, TEX41, PRSS57]  \n",
       "13                                                 []  \n",
       "14                                                 []  \n",
       "15                                                 []  \n",
       "16                                                 []  \n",
       "17                                                 []  \n",
       "18  [AC109829.1, ADH7, KLB, NPIPB8, SULT1A1, CCDC1...  \n",
       "19  [MYH6, CMTM5, IL25, EFS, KIAA1755, MYH7, SLC22...  \n",
       "20  [NPR3, C11orf89, LSP1, CACNB2, ATP2B1, TNNI2, ...  \n",
       "21                                                 []  \n",
       "22                                                 []  \n",
       "23  [HLA-DQB1, XXbac-BPG254F23.7, HLA-DQA1, HLA-DQ...  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_diseases_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_diseases_all.to_csv('/SSD/home/sofia/retina-phenotypes/complementary/GWAS_postprocessing/'+neale_csv_name, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2- Neale Pascal genes AND GWAS catalog"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### a) Select the diseases of interest on list_pheno_neale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Read the previous df\n",
    "df_genes_neale= pd.read_csv('/SSD/home/sofia/retina-phenotypes/complementary/GWAS_postprocessing/'+neale_csv_name, converters={'names': pd.eval})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pheno</th>\n",
       "      <th>name_pheno</th>\n",
       "      <th>N genes</th>\n",
       "      <th>names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>only_rs_notna_mod_variants_4080_irnt</td>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>959</td>\n",
       "      <td>[NPPA, NPPB, MTHFR, CLCN6, C1orf167, NPR3, AGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  pheno name_pheno  N genes  \\\n",
       "2  only_rs_notna_mod_variants_4080_irnt  4080_irnt      959   \n",
       "\n",
       "                                               names  \n",
       "2  [NPPA, NPPB, MTHFR, CLCN6, C1orf167, NPR3, AGT...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_genes_neale['liststring'] = [''.join(map(str, l)) for l in df_genes_neale['names']]\n",
    "#df_genes_neale['names'].iloc[4][2]\n",
    "\n",
    "list_pheno_neale=['4080_irnt'] #list(df_genes_neale['pheno'])\n",
    "#['4080_irnt', '102_irnt', '4056_irnt', '5610_3', '4079_irnt', '1717', '3894_irnt']\n",
    "\n",
    "df_genes_neale_red = df_genes_neale[df_genes_neale[\"name_pheno\"].isin(list_pheno_neale)]\n",
    "#df_genes_neale_red['pheno_name'] = ['SBP_irnt', 'PR_irnt', 'Stroke_irnt', 'Presbyopia_3', 'DBP_irnt', 'Skin_color', 'heart_attack_irnt']\n",
    "#df_genes_neale_red = df_genes_neale.copy()\n",
    "#df_genes_neale_red\n",
    "df_genes_neale_red"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Select the file of interest:\n",
    "#df_genes_neale_red['names'][df_genes_neale_red['pheno']=='only_rs_notna_mod_variants_4079_irnt']\n",
    "#set_only_rs_notna_mod_variants_4079_irnt = set(df_genes_neale_red['names'].iloc[5])\n",
    "#set_only_rs_notna_mod_variants_4079_raw = set(df_genes_neale_red['names'].iloc[6])\n",
    "#print(len(set_only_rs_notna_mod_variants_4079_irnt), len(set_only_rs_notna_mod_variants_4079_raw))\n",
    "#print('Intersection of both: ', len(set_only_rs_notna_mod_variants_4079_irnt&set_only_rs_notna_mod_variants_4079_raw))\n",
    "#df_genes_neale_red.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### b) GWAS catalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_catalog = '/HDD/data/GWAS_catalog/'\n",
    "os.chdir(dir_catalog)\n",
    "\n",
    "l_genes_shared=[]\n",
    "for file_name in glob.glob(\"gwas-association-downloaded_SBP.tsv\"): #glob(\"*.tsv\"):\n",
    "    #print(file_name)\n",
    "    df_gwas = pd.read_table(dir_catalog+ str(file_name))\n",
    "    #df_reduced = df_gwas[['MAPPED_TRAIT', 'REPORTED GENE(S)', 'MAPPED_GENE']]\n",
    "    #print(df_reduced.head(5))\n",
    "    for i in range(len(df_gwas)):\n",
    "        gen = df_gwas['REPORTED GENE(S)'].iloc[i]\n",
    "        for j in range(len(df_genes_neale_red)):\n",
    "            for m in range(len(df_genes_neale_red['names'].iloc[j])):\n",
    "                if df_genes_neale_red['names'].iloc[j][m]==gen:\n",
    "                    data={\n",
    "                    'Neale_pheno':  df_genes_neale_red['name_pheno'].iloc[j],\n",
    "                    'MAPPED_TRAIT': df_gwas['MAPPED_TRAIT'].iloc[i],\n",
    "                    'gene': df_genes_neale_red['names'].iloc[j][m],\n",
    "                    'gene1': gen,\n",
    "                    'INITIAL SAMPLE SIZE': df_gwas['INITIAL SAMPLE SIZE'].iloc[i]\n",
    "                    }\n",
    "                    l_genes_shared.append(data)\n",
    "\n",
    "df_genes_shared =pd.DataFrame(l_genes_shared)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Neale_pheno</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene1</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>ARNTL</td>\n",
       "      <td>ARNTL</td>\n",
       "      <td>291,107 British ancestry individuals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>NMT1</td>\n",
       "      <td>NMT1</td>\n",
       "      <td>291,107 British ancestry individuals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>NR1H3</td>\n",
       "      <td>NR1H3</td>\n",
       "      <td>291,107 British ancestry individuals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>GIP</td>\n",
       "      <td>GIP</td>\n",
       "      <td>291,107 British ancestry individuals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>USP36</td>\n",
       "      <td>USP36</td>\n",
       "      <td>291,107 British ancestry individuals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>RP11-1055B8.6</td>\n",
       "      <td>RP11-1055B8.6</td>\n",
       "      <td>757,601 European ancestry individuals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>757,601 European ancestry individuals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>POM121C</td>\n",
       "      <td>POM121C</td>\n",
       "      <td>757,601 European ancestry individuals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>THSD7B</td>\n",
       "      <td>THSD7B</td>\n",
       "      <td>757,601 European ancestry individuals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>4080_irnt</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>KDM4B</td>\n",
       "      <td>KDM4B</td>\n",
       "      <td>757,601 European ancestry individuals</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Neale_pheno             MAPPED_TRAIT           gene          gene1  \\\n",
       "0     4080_irnt  systolic blood pressure          ARNTL          ARNTL   \n",
       "1     4080_irnt  systolic blood pressure           NMT1           NMT1   \n",
       "2     4080_irnt  systolic blood pressure          NR1H3          NR1H3   \n",
       "3     4080_irnt  systolic blood pressure            GIP            GIP   \n",
       "4     4080_irnt  systolic blood pressure          USP36          USP36   \n",
       "..          ...                      ...            ...            ...   \n",
       "200   4080_irnt  systolic blood pressure  RP11-1055B8.6  RP11-1055B8.6   \n",
       "201   4080_irnt  systolic blood pressure         CTNNB1         CTNNB1   \n",
       "202   4080_irnt  systolic blood pressure        POM121C        POM121C   \n",
       "203   4080_irnt  systolic blood pressure         THSD7B         THSD7B   \n",
       "204   4080_irnt  systolic blood pressure          KDM4B          KDM4B   \n",
       "\n",
       "                       INITIAL SAMPLE SIZE  \n",
       "0     291,107 British ancestry individuals  \n",
       "1     291,107 British ancestry individuals  \n",
       "2     291,107 British ancestry individuals  \n",
       "3     291,107 British ancestry individuals  \n",
       "4     291,107 British ancestry individuals  \n",
       "..                                     ...  \n",
       "200  757,601 European ancestry individuals  \n",
       "201  757,601 European ancestry individuals  \n",
       "202  757,601 European ancestry individuals  \n",
       "203  757,601 European ancestry individuals  \n",
       "204  757,601 European ancestry individuals  \n",
       "\n",
       "[205 rows x 5 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Drop genes duplicated\n",
    "df_genes_shared_no_dup = df_genes_shared.drop_duplicates(subset = [\"gene\"])\n",
    "df_genes_shared_no_dup.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "### IMPORTANT. NOT FILTERED BY ETNICITY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_genes_shared.to_csv('/SSD/home/sofia/retina-phenotypes/complementary/GWAS_postprocessing/5_09_2022_shared_diseases_genes.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OLD VERSION (relevant -> create a list with all the genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_diseases=pd.read_csv('/SSD/home/sofia/retina-phenotypes/complementary/GWAS_postprocessing/1_09_2022_diseases_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Example:\n",
    "print(len(df_diseases['names'].iloc[4]))\n",
    "#### Do your strings look like this: “[‘strawberry’, ‘apple’, ‘orange’]”?\n",
    "# In that case, there is a quick method using the apply() and eval() functions.\n",
    "df_diseases['names'] = df_diseases['names'].apply(eval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create a list with all the genes\n",
    "list_genes_diseases_pascal=[]\n",
    "for i in range(len(df_diseases)):\n",
    "    if len(df_diseases['names'].iloc[i])==0: \n",
    "        continue\n",
    "        #print(df_diseases['names'].iloc[i])\n",
    "    else:\n",
    "        list_genes_diseases_pascal = list_genes_diseases_pascal + df_diseases['names'].iloc[i]\n",
    "        #list_genes_diseases_pascal.append(list(df_diseases['names'].iloc[i]))\n",
    "#list_genes_diseases_pascal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### N* -> OLD VERSION (does not work)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### NEED TO ADD df_genes_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gen in list_genes_diseases_pascal:\n",
    "    if df_genes_count[df_genes_count['Unnamed: 0']==gen].empty:\n",
    "        continue\n",
    "    else:\n",
    "        print(gen)\n",
    "        print(df_genes_count[df_genes_count['Unnamed: 0']==gen], '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### OLD VERSION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(dir_catalog)\n",
    "\n",
    "for file_name in glob.glob(\"*.tsv\"):\n",
    "    #print(file_name)\n",
    "    df_gwas = pd.read_table(dir_catalog+ str(file_name))\n",
    "    #df_reduced = df_gwas[['MAPPED_TRAIT', 'REPORTED GENE(S)', 'MAPPED_GENE']]\n",
    "    #print(df_reduced.head(5))\n",
    "    for i in range(len(df_gwas)):\n",
    "        gen = df_gwas['REPORTED GENE(S)'].iloc[i]\n",
    "        if df_genes_count[df_genes_count['Unnamed: 0']==gen].empty:\n",
    "            continue\n",
    "        else:\n",
    "            print(df_gwas['MAPPED_TRAIT'].iloc[i], df_gwas['INITIAL SAMPLE SIZE'].iloc[i])\n",
    "            print(df_genes_count[df_genes_count['Unnamed: 0']==gen])\n",
    "            print('\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_genes_count[df_genes_count['Unnamed: 0']=='CTD-2008P7.9']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
